topic_modeling/scrape_training_data/get_article_links.py

```diff
@@ -5,50 +5,78 @@
 from IPython.core.display import clear_output
 from time import time
 
-# change range to 0,50 when using cluster
-pages = list(map(str, range(4)))
 
-# change range to 2012,2019 when using cluster
-year_urls = list(range(2015, 2016))
+def get_article_links(start_year, end_year, num_pages):
 
-article_links = []
+    """Collect and save a list of links that correspond to
+    articles from The Dartmouth daily newspaper.
 
-start_time = time()
-requests = 0
+    Parameters
+    ----------
+    start_year : int
+        Year in the archives to start looking for articles
+    end_year : int
+        Year in the archives to stop looking for articles
+    num_pages : int
+        Number of archive pages that fit each year range to
+        inspect
 
-for year_url in year_urls:
+    Returns
+    -------
+    text file
+        A text file ('article_links.txt' that contains the
+        link to every article that appears in specificed
+        year and page range.
 
-    for page in pages:
+    """
 
-        html_page = urllib.request.urlopen(
-            "http://www.thedartmouth.com/search?q=0&page={}&"
-            "ti=0&tg=0&ty=0&ts_month=0&ts_day=0&ts_year={}&"
-            "te_month=0&te_day=0&te_year={}&s=0&au=0&o=0&"
-            "a=1".format(page, year_url, year_url+1))
+    # change range to 0,50 when using cluster
+    pages = list(map(str, range(num_pages)))
 
-        # Pause the loop
-        sleep(randint(8, 15))
+    # change range to 2012,2019 when using cluster
+    year_urls = list(range(start_year, end_year))
 
-        # Monitor the requests
-        requests += 1
-        elapsed_time = time() - start_time
-        print('Request:{}; Frequency: {} requests/s'.format(
-            requests, requests / elapsed_time))
-        clear_output(wait=True)
+    article_links = []
 
-        soup = BeautifulSoup(html_page, 'html.parser')
+    start_time = time()
+    requests = 0
 
-        links = []
-        for link in soup.find_all('a'):
-            links.append(link.get('href'))
+    for year_url in year_urls:
 
-        for link in links:
-            if link not in article_links:
-                if 'article' in link.split('/'):
-                    article_links.append(link)
+        for page in pages:
 
+            html_page = urllib.request.urlopen(
+                "http://www.thedartmouth.com/search?q=0&page={}&"
+                "ti=0&tg=0&ty=0&ts_month=0&ts_day=0&ts_year={}&"
+                "te_month=0&te_day=0&te_year={}&s=0&au=0&o=0&"
+                "a=1".format(page, year_url, year_url+1))
```

Comment:
with such long strings, best to rely not on positional arguments for the `format()` but specify them as keyword arguments and use then `{page}`, `{year}`, `{next_year}` - so that url would be easy to read and thus avoid possible mistakes when you have to edit it

Reply:
eventually, if you need to scrape some other bigger site, or even now if it takes too long, you might like to learn/use https://scrapy.org which could crawl in parallel etc